rfcs/0000-light-rail-trip-model.md

```diff
@@ -0,0 +1,211 @@
+
+## Glides schedule import
+
+Glides will import schedule data from either its own HASTUS export or, in the future, TODS. In either case, trip data is loaded into the `scheduled_trips` table. The HASTUS trip ID will be used directly to populate the `trip_id` column. While working directly with HASTUS, some modifications will be necessary to which IDs we actually use. Currently, our export gets a column called `trp_number`, but the actual, database-unique ID is called `trp_int_number`, and the latter is also the value published by `gtfs_creator`. The significant different here is that cars grouped into one logical trip all have the same `trp_number` generally, but as `trp_int_number` is database-unique, that value will differ. We want the benefits of both combined trip information for multiple cars as well as uniqueness and agreement between systems. To achieve this, we will combine the trips for the different constituent cars and determine the ultimate ID in the following manner:
+
+1. Group the trip entries for each schedule ID by the `trp_number` value.
+1. For each resulting group, take the trip record for the lead car and use its `trp_int_number`.
+1. Use information about the trailer trip for data like the run number, but discard its `trp_int_number`.
+
+## Glides trainsheet edit events
+
+The `com.mbta.ctd.glides.trips_updated` event schema will receive a version bump, modifying the `TripKey` datatype. `TripKey` will be the same for both added and scheduled trips, in either case being represented as an object with `"serviceDate"` and `"tripId"` keys, where the `tripId` is sourced from the `trip_id` column of `scheduled_trips` for a scheduled trip, or the auto-generated Glides ID for an added trip.
+
+## Glides train - trip assignment events
+
+A new `com.mbta.ctd.glides.vehicle_trip_assignment` event will be established. The exact schema will be defined via the pull request review procedure on the [`mbta/schemas`](https://github.com/mbta/schemas/) repository, but there are some basic criteria that can be established. Events will at least have a `tripId` and `vehicleId` key. The `tripId` will be the scheduled or added trip ID from Glides, and the `vehicleId` will correspond to the ID of the vehicle in the GTFS-rt feed. Each event will represent the assignment of the given `vehicleId` to the given `tripId`. Either a variation of this event or a separate event will exist for unassigning a vehicle from a trip.
+
+The Glides application will be primarily responsible for establishing the train - trip matching. While exact details of the logic are beyond the scope of this RFC, factors that are likely to be weighed include:
+
+1. Cars entered for trips on trainsheets
```

Comment:
Does Glides have the information needed to do this? Currently RTR does this by (slightly oversimplified)

1. Parsing the OCS provided [route number](https://github.com/mbta/rtr/blob/master/docs/ocs_messages.md?plain=1#L98) to [determine the terminals](https://github.com/mbta/rtr/blob/master/lib/rtr/patterns/pattern.ex#L74)
2. Using the vehicle's current position to determine what track it's on and therefore, what direction it should be moving in.
3. Use 1 and 2 combined with the current time to guess the GTFS trip it's on (or assign an added trip if we can't match)

Does Glides have enough knowledge of the tracks to do 2? Or is the idea that in the future we'll be getting reliable enough information from operators that we don't have to infer, they'll tell us when assigning a train to a new trip?

Also how should reverse trips be handled?

Reply:
Sorry, pedantic but important point, but operators do not input any data into Glides; inspectors do.

Regarding 1, this does mean that Glides will need an internal data feed that includes at least the AVI code information, which we don't currently have. I know there has been some background discussion of how to handle this sort of internal information in our feeds but I will add a note explicitly calling that out in the "Glides data requirements" section below (dcebbf1).

I do agree that 2 is the trickiest part here, and the one where RTR arguably has the biggest advantage over Glides currently. I'll have to think the implications of that over a bit more and get back to you.

Reply:
> Also how should reverse trips be handled?

This logic is only about determining current trip assignments; I think figuring out what trip a train will be doing next is still a question of predictions, and one where looking at the schedule can come into play (looking at runs or blocks to determine what a train will do next).

I suppose one important case here is where RTR currently generates an ADDED trip, it will also create a reverse trip if the end station of the ADDED trip is among a certain set of terminals, right? That might be a little bit trickier in this approach, but we might also be able to avoid having as many ADDED trips.

Reply:
I think you're right that [my reply in the other thread](https://github.com/mbta/technology-docs/pull/25#discussion_r1704540308) answers most of this question.

One more pedantic point that Government Center and BC are swapped in your example of which one has a user entering data in Glides and which one gets reverse predictions from RTR, but other than that it's the right basic idea.

I'm thinking at this from a few different angles now and I think it will actually be useful to break this down into a few different cases:

1. A regularly scheduled trip on the Glides trainsheet, with car numbers entered by the inspector and then also transferred over to act as the car numbers for the return trip by Glides. This is the most common and also easiest case.
2. A trip that isn't in the Glides schedule but is explicitly added to the trainsheet by a Glides user. These can be either one way or two way. If it's a two-way trip it should work just like case 1. If it's a one-way added trip we won't automatically get a trip assignment once the train turns around, but maybe that's actually appropriate if the user only indicated that the trip is one-way, and we should instead wait for the user to explicitly add another train for that consist?
3. A trip that never gets added to the Glides trainsheet at all. This is definitely the worst case in a lot of ways. Part of me wants to say that this should be uncommon enough that it might not be worth handling specially. Our team is also going to be doing some work soon on understanding how often there are trains running around that don't correspond to anything on the Glides trainsheets at all, which should give us a sense of how big (or not) of a problem this is, which might change my mind.

Reply:
> I do agree that 2 is the trickiest part here, and the one where RTR arguably has the biggest advantage over Glides currently. I'll have to think the implications of that over a bit more and get back to you.

I did want to quickly circle back to say that I definitely haven't forgotten about this. I did at least make sure to note this as an important consideration in the body of the document (361ad05). I may decide to make other changes but I want to make sure that the concern is at the very least noted.

Reply:
Okay, based on a conversation with @skyqrose I have written up some more thoughts on this matter - long story short I think it's okay, but let me know what you think @Whoops. 2fffccb

Reply:
probably a question for @Whoops: how is the next trip/direction ID in the GTFS-RT feed calculated now? If it depends on the trip assignment, then I'm not sure how this would work for Glides.

Reply:
It's based on pattern, and I addressed how that is determined in the content of the changes I just pushed.